Queuing and batching support in java servlet.

praves77 opened the pull request:
1. Added a new class BatchProcessor to manage queue and periodic taskRunner to send events in batches to moesif.
2. Added two timer tasks for repeated fixed delay execution. One tasks pulls app config periodically from the server and update the local cache. Second tasks runs the batch processor's method periodically to send the events to the server.
3. Added sanity in add batch event to check liveness of the task-runner. If batch event taskRunner is not active then restart it.
4. Did few refactoring to isolate the functionality based on name of the methods.
5. Bumped up the version of the servlet to 1.6.17 and updated the other affected README & repos.

Ticket link: https://www.pivotaltracker.com/n/projects/1646401/stories/172421968

praves77 commented:
@dgilling I prefixed the configuration parameter with `event` (like `eventBatchSize` ) . I can rename the updateConfigTimeInMin to `updateAppConfigTimeInMin` or everything in seconds. This is used to auto pull appConfig and update the local cache. (nice recommendation by @keyur9 to use async call).

praves77 commented:
There are few magic numbers in the code (which I don't like but too many config parameters make the usability difficult) like (how long to wait to restart timer job, if last executed task was sometime back - for 5 secs timer, I used 5*60 secs to wait before restarting the timer).

dgilling commented on moesif-servlet/src/main/java/com/moesif/servlet/MoesifFilter.java:
```diff
@@ -427,6 +444,143 @@ private EventResponseModel getEventResponseModel(LoggingHttpServletResponseWrapp
+  private void rescheduleSendEventsJobIfNeeded() {
+
+    // Check if batchJob is already running then return
+    if (this.batchProcessor.isJobRunning()) {
```
Great to see not spawning too many threads but also verifying events not stuck forever with auto reschedule

dgilling commented on moesif-servlet/src/main/java/com/moesif/servlet/MoesifFilter.java:
```diff
@@ -427,6 +444,143 @@ private EventResponseModel getEventResponseModel(LoggingHttpServletResponseWrapp
+  /**
+   * Method to create scheduled job for updating config periodically.
+   */
+  private void scheduleAppConfigJob() {
```
Future change: 5 mins is prob fine for appConfig, but might recommend 1 min or 3 mins in future change once this SDK supports governance rules. Another option is to store Etag and schedule as needed (Etag change or 5 mins, whichever soonest)

praves77 replied:
@dgilling For better performance, the code has been changed to async API call in fire and forget mode because volume of such calls will be high. that means we are not waiting for response/Etag. That's why a timer based jobRunner was added to pull/update appConfig periodically.
I will add support in callback handler and see how it performs.

praves77 replied:
Done

dgilling commented on moesif-servlet/src/main/java/com/moesif/servlet/MoesifFilter.java:
```diff
@@ -427,6 +444,143 @@ private EventResponseModel getEventResponseModel(LoggingHttpServletResponseWrapp
+
+      // Check send batchEvent job periodically based on counter if rescheduling is needed.
+      if (sendBatchJobALiveCounter > 100) {
+        this.rescheduleSendEventsJobIfNeeded();
```
Can we print debug message here. Helpful for perf tweaks.

dgilling reviewed (changes requested):
Looks good! Just some small tweaks requested and update README.md

praves77 commented:
Following scenarios were tested : 
1. Don't add any event. taskRunner was returning immediately.
2. Add smaller than batchSize events. taskRunner sent those events in a single batch during periodic run.
3. Add many events (> batchSize) quickly, multiple batches were sent by the taskRunner
4. Changed the appConfig and found that app config was automatically updated after taskRunner for autoUpdateAppConfig executed. Updated appConfig was applied in the next batch events.
5. Tested with Apache Jmeter to fire multiple events and ran it for 10 mins.
6. Fake userName and companyNames were used and all events showed up in the Moesif 

Let me know if I missed any.

dgilling commented:
@praves77 , I would recommend also with typical payload (<1KB JSON) and also large payload (500KB). Also, what about a connection timeout or error. Does it retry, how many times? @keyur9 , might have some tests hady.

keyur9 commented:
Will share some more tests with @praves77 offline.

CodeYouMust commented on moesif-servlet/src/main/java/com/moesif/servlet/BatchProcessor.java:
```diff
@@ -0,0 +1,150 @@
+                try {
+                    this.moesifApi.getAPI().createEventsBatchAsync(curEventList, callBack);
+                } catch(JsonProcessingException e) {
+                    logger.warning("Failed to send batch events. " + e.toString());
```
Could there be other types of exceptions here? Also if there is a batch of events, and if a single event causes parsing exception, will entire batch fail?

CodeYouMust commented on moesif-servlet/src/main/java/com/moesif/servlet/BatchProcessor.java:
```diff
@@ -0,0 +1,150 @@
+        } catch (InterruptedException e) {
+            logger.warning("Add event failed. " + e.toString());
+        } catch (Exception e) {
+            logger.warning("Add event failed. " + e.toString());
```
minor nitpic: since we perform same action, do we need two catch here?

praves77 replied:
Thanks @CodeYouMust for checking. No, we do not. I added it to keep IntelliJ happy (probably suggested based on their linting rule). Will remove it.

CodeYouMust commented on moesif-servlet/src/main/java/com/moesif/servlet/MoesifFilter.java:
```diff
@@ -427,6 +444,143 @@ private EventResponseModel getEventResponseModel(LoggingHttpServletResponseWrapp
+    // Check if we need to reschedule job to send batch events
+    // if the last job runtime is more than 5 minutes.
+    final long MAX_TARDINESS_SEND_EVENT_JOB = this.config.batchMaxTimeInSec * 60;      // in seconds
+    final long diff = new Date().getTime() - this.batchProcessor.scheduledExecutionTime();
```
this should be safe for possible clock adjustments such as daylight savings time change, i think?

praves77 replied:
Good question @CodeYouMust I assumed server runs in UTC zone and locale is handled by client to avoid TZ pitfalls. Please correct me if I am wrong.

CodeYouMust replied:
@praves77 As far as server tz, since this is public sdk, we cant assume UTC. but I think since this is long time it should be tz agnostic, just rusty on it, so should confirm in all places where we check time. if system time changes as a result of clock sync change (eg ntp sync), then i dunno, but that should be very unlikely that it changes more than several seconds. Just maybe defensively ensure we dont have `long diff` beyond certain positive number range.

CodeYouMust commented on moesif-servlet/src/main/java/com/moesif/servlet/BatchProcessor.java:
```diff
@@ -0,0 +1,150 @@
+            }
+
+            public void onFailure(HttpContext context, Throwable error) {
+                // ignore
```
im not too familiar with this but, for `onSuccess` it may still have status code that is 4xx ot 5xx? [See sdk link](https://github.com/Moesif/moesifapi-java#2a-send-the-events-batch-asynchronously) and may need to treat them as failure? What would be the behavior with token that fails auth all the time vs intermittent network failure?
```
assertEquals("Status is not 201",
                201, httpContext.getResponse().getStatusCode());
```

praves77 replied:
Done

praves77 commented:
> 

@dgilling I thought that retry logic anf gzip might in API handler - common to all. I am adding 3 retries.

praves77 commented on moesif-servlet/pom.xml:
```diff
@@ -3,7 +3,7 @@
   <groupId>com.moesif.servlet</groupId>
   <artifactId>moesif-servlet</artifactId>
-  <version>1.6.16</version>
+  <version>1.6.17</version>
```
Done

praves77 commented:
Updated PR based on suggestions + added support for retry.
